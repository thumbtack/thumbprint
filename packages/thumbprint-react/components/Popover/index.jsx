import React, { useState } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { Manager, Reference, Popper } from 'react-popper';
import startsWith from 'lodash/startsWith';

import * as tokens from '@thumbtack/thumbprint-tokens';

import ConditionalPortal from '../../utils/ConditionalPortal.jsx';
import canUseDOM from '../../utils/can-use-dom';
import useCloseOnEscape from '../../utils/use-close-on-escape';
import useFocusTrap from '../../utils/use-focus-trap';

import { TextButton } from '../Button/index.jsx';
import { Themed } from '../UIAction/index.jsx';
import { NavigationCloseTiny } from '../../icons/index.jsx';

import styles from './index.module.scss';

const Popover = ({
    children,
    launcher,
    position,
    isOpen,
    onCloseClick,
    container,
    accessibilityLabel,
}) => {
    // Appends the tooltip right before `</body>` when true. Used to prevent z-index and positioning
    // issues.
    const shouldDisplace = container === 'body';

    // Using `useState` instead of `useRef `to allow multiple refs. See Image for another example.
    const [wrapperEl, setWrapperEl] = useState(null);

    const shouldTrapFocus = canUseDOM && wrapperEl;
    const shouldBindEscListener = canUseDOM && isOpen;

    useCloseOnEscape(onCloseClick, shouldBindEscListener);
    useFocusTrap(wrapperEl, shouldTrapFocus, {
        clickOutsideDeactivates: true,
        // Set initial focus to the modal wrapper itself instead of focusing on the first
        // focusable element by default.
        initialFocus: wrapperEl,
    });

    return (
        <Manager>
            <Reference>{({ ref }) => launcher({ ref })}</Reference>
            <ConditionalPortal shouldDisplace={shouldDisplace}>
                {canUseDOM && (
                    <Popper
                        placement={position}
                        modifiers={{
                            offset: { offset: `0, ${tokens.tpSpace3}` },
                            preventOverflow: { boundariesElement: 'window' },
                        }}
                        positionFixed={false}
                    >
                        {({ ref: popperRef, style, placement, arrowProps }) => (
                            // Use tabIndex="-1" to allow programmatic focus (as initialFocus node
                            // for focus-trap) but not be tabbable by user.
                            <div
                                role="dialog"
                                aria-label={accessibilityLabel}
                                tabIndex="-1"
                                ref={el => {
                                    setWrapperEl(el);
                                    popperRef(el);
                                }}
                                className={classNames({
                                    [styles.root]: true,
                                    [styles.open]: isOpen,
                                })}
                                style={style}
                                data-placement={placement}
                            >
                                {children}

                                <div className={styles.closeButton}>
                                    <TextButton
                                        accessibilityLabel="Close popover"
                                        iconLeft={
                                            <NavigationCloseTiny
                                                className={styles.closeButtonIcon}
                                            />
                                        }
                                        theme="inherit"
                                        onClick={onCloseClick}
                                    />
                                </div>

                                <div
                                    className={classNames({
                                        [styles.nubbin]: true,
                                        [styles.nubbinTop]: startsWith(placement, 'bottom'),
                                        [styles.nubbinBottom]: startsWith(placement, 'top'),
                                        [styles.nubbinLeft]: startsWith(placement, 'right'),
                                        [styles.nubbinRight]: startsWith(placement, 'left'),
                                    })}
                                    ref={arrowProps.ref}
                                    style={arrowProps.style}
                                />
                            </div>
                        )}
                    </Popper>
                )}
            </ConditionalPortal>
        </Manager>
    );
};

Popover.propTypes = {
    /**
     * Contents for the Popover. Usually a `PopoverTitle`, `PopoverBody`, and `PopoverPrimaryButton`
     */
    children: PropTypes.node.isRequired,
    /**
     * A function that renders JSX and receives an object with `ref`.
     * All of these props must be added to the component within the render prop.
     */
    launcher: PropTypes.func.isRequired,
    /**
     * Position of popover relative to the launcher.
     */
    position: PropTypes.oneOf([
        'top-start',
        'top',
        'top-end',
        'bottom-start',
        'bottom',
        'bottom-end',
        'left-start',
        'left',
        'left-end',
        'right-start',
        'right',
        'right-end',
    ]),
    /**
     * Whether or not the popover is visible.
     */
    isOpen: PropTypes.bool,
    /**
     * Function called when the close button is clicked. You should cause this to set `isOpen=false`
     * in your parent component.
     */
    onCloseClick: PropTypes.func.isRequired,
    /**
     * By default popovers will render right before the `</body>` tag.
     * Setting the `container` to `inline` causes the tooltip to remain where it was added to the
     * DOM.
     * This option is helpful to work around z-index and positioning issues.
     */
    container: PropTypes.oneOf(['inline', 'body']),
    /**
     * Accessibility title used to describe the content of the popover to screen readers.
     */
    accessibilityLabel: PropTypes.string,
};

Popover.defaultProps = {
    isOpen: false,
    position: 'top',
    container: 'body',
    accessibilityLabel: 'Popover',
};

export default Popover;

// === Sub-components ===

const PopoverTitle = ({ children }) => <div className={styles.popoverTitle}>{children}</div>;

PopoverTitle.propTypes = {
    /**
     * A title for the popover.
     */
    children: PropTypes.string.isRequired,
};

const PopoverBody = ({ children }) => <div className={styles.popoverBody}>{children}</div>;

PopoverBody.propTypes = {
    /**
     * The body description text for the popover.
     */
    children: PropTypes.node.isRequired,
};

const PopoverPrimaryButton = ({ children, onClick }) => (
    <Themed size="small" onClick={onClick} theme="popover-primary">
        {children}
    </Themed>
);

PopoverPrimaryButton.propTypes = {
    /**
     * The text for this button.
     */
    children: PropTypes.string.isRequired,
    /**
     * Function called when the button is clicked.
     */
    onClick: PropTypes.func.isRequired,
};

const PopoverSecondaryButton = ({ children, onClick, to }) => (
    <Themed size="small" onClick={onClick} to={to} theme="popover-secondary" shouldOpenInNewTab>
        {children}
    </Themed>
);

PopoverSecondaryButton.propTypes = {
    /**
     * The text for this button.
     */
    children: PropTypes.string.isRequired,
    /**
     * Function called when the button is clicked, or link to visit.
     */
    onClick: PropTypes.func,
    /**
     * Link to visit when the button is clicked. It will be opened in a new tab.
     */
    to: PropTypes.string,
};

PopoverSecondaryButton.defaultProps = {
    onClick: undefined,
    to: undefined,
};

export { PopoverTitle, PopoverBody, PopoverPrimaryButton, PopoverSecondaryButton };
