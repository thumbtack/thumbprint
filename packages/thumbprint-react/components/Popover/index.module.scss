@import '~@thumbtack/thumbprint-tokens/dist/scss/_index';
@import '~@thumbtack/thumbprint-scss/mixins';

.root {
    background: $tp-color__white;
    box-shadow: 0px 2px 10px rgba(0, 0, 0, 0.2);
    padding: $tp-space__3 $tp-space__4;
    border-radius: $tp-border-radius__base;
    border-top: 4px solid $tp-color__indigo;
    position: relative;
    max-width: 300px;
    visibility: hidden;
    opacity: 0;
    outline: none;
}

.open {
    visibility: visible;
    opacity: 1;
}

.popoverTitle {
    font-size: $tp-font__title__7__size;
    line-height: $tp-font__title__7__line-height;
    font-weight: $tp-font__title__7__weight;
    color: $tp-color__black;
    margin-bottom: $tp-space__2;
}

.popoverBody {
    font-size: $tp-font__body__2__size;
    line-height: $tp-font__body__2__line-height;
    color: $tp-color__black;
    margin-bottom: $tp-space__3;
}

.closeButton {
    position: absolute;
    top: 0;
    right: 0;

    &Icon {
        // Increase click target of the button to make it easier to tap.
        margin: $tp-space__2;
    }
}

.nubbin {
    width: 0;
    height: 0;
    border-style: solid;
    position: absolute;
    margin: $tp-space__2;
    // Create an arrow with CSS.
    border-width: $tp-space__2 $tp-space__2 0 $tp-space__2;
    border-left-color: transparent;
    border-right-color: transparent;
    border-bottom-color: transparent;
    margin-top: 0;
    margin-bottom: 0;

    &Top {
        transform: scale(-1);
        top: -12px;
        border-top-color: $tp-color__indigo;
    }

    &Bottom {
        transform: scale(1);
        bottom: -8px;
        border-top-color: $tp-color__white;
    }

    &Left {
        transform: rotate(90deg);
        left: -20px;
        border-top-color: $tp-color__white;
    }

    &Right {
        transform: rotate(-90deg);
        right: -20px;
        border-top-color: $tp-color__white;
    }
}
