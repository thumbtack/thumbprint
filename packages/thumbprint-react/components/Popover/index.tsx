import React, { useState } from 'react';
import classNames from 'classnames';
import { Manager, Reference, Popper, RefHandler } from 'react-popper';
import startsWith from 'lodash/startsWith';

import * as tokens from '@thumbtack/thumbprint-tokens';

import ConditionalPortal from '../../utils/ConditionalPortal';
import canUseDOM from '../../utils/can-use-dom';
import useCloseOnEscape from '../../utils/use-close-on-escape';
import useFocusTrap from '../../utils/use-focus-trap';

import { TextButton } from '../Button/index';
import { Themed } from '../UIAction/index';

import styles from './index.module.scss';

interface PopoverPropTypes {
    /**
     * Contents for the Popover. Usually a `PopoverTitle`, `PopoverBody`, and `PopoverPrimaryButton`
     */
    children: React.ReactNode;
    /**
     * A function that renders JSX and receives an object with `ref`.
     * All of these props must be added to the component within the render prop.
     */
    launcher: ({ ref }: { ref: RefHandler }) => React.ReactNode;
    /**
     * Position of popover relative to the launcher.
     */
    position?:
        | 'top-start'
        | 'top'
        | 'top-end'
        | 'bottom-start'
        | 'bottom'
        | 'bottom-end'
        | 'left-start'
        | 'left'
        | 'left-end'
        | 'right-start'
        | 'right'
        | 'right-end';
    /**
     * Whether or not the popover is visible.
     */
    isOpen?: boolean;
    /**
     * Function called when the close button is clicked. You should cause this to set `isOpen=false`
     * in your parent component.
     */
    onCloseClick: () => void;
    /**
     * By default popovers will render right before the `</body>` tag.
     * Setting the `container` to `inline` causes the tooltip to remain where it was added to the
     * DOM.
     * This option is helpful to work around z-index and positioning issues.
     */
    container?: 'inline' | 'body';
    /**
     * Accessibility title used to describe the content of the popover to screen readers.
     */
    accessibilityLabel?: string;
}

export default function Popover({
    children,
    launcher,
    onCloseClick,
    position = 'top',
    isOpen = false,
    container = 'body',
    accessibilityLabel = 'Popover',
}: PopoverPropTypes): JSX.Element {
    // Appends the tooltip right before `</body>` when true. Used to prevent z-index and positioning
    // issues.
    const shouldDisplace = container === 'body';

    // Using `useState` instead of `useRef `to allow multiple refs. See Image for another example.
    const [wrapperEl, setWrapperEl] = useState<HTMLDivElement | null>(null);

    const shouldTrapFocus: boolean = canUseDOM && !!wrapperEl;
    const shouldBindEscListener: boolean = canUseDOM && isOpen;

    useCloseOnEscape(onCloseClick, shouldBindEscListener);
    useFocusTrap(wrapperEl, shouldTrapFocus, wrapperEl);

    return (
        <Manager>
            <Reference>{({ ref }): React.ReactNode => launcher({ ref })}</Reference>
            <ConditionalPortal shouldDisplace={shouldDisplace}>
                {canUseDOM && (
                    <Popper
                        placement={position}
                        modifiers={{
                            offset: { offset: `0, ${tokens.tpSpace3}` },
                            preventOverflow: { boundariesElement: 'window' },
                        }}
                        positionFixed={false}
                    >
                        {({ ref: popperRef, style, placement, arrowProps }): JSX.Element => (
                            // Use tabIndex={-1} to allow programmatic focus (as initialFocus node
                            // for focus-trap) but not be tabbable by user.
                            <div
                                role="dialog"
                                aria-label={accessibilityLabel}
                                tabIndex={-1}
                                ref={(el: HTMLDivElement | null): void => {
                                    setWrapperEl(el);
                                    popperRef(el);
                                }}
                                className={classNames({
                                    [styles.root]: true,
                                    [styles.open]: isOpen,
                                })}
                                style={style}
                                data-placement={placement}
                            >
                                {children}

                                <div className={styles.closeButton}>
                                    <TextButton
                                        accessibilityLabel="Close popover"
                                        iconLeft={
                                            <svg
                                                viewBox="0 0 24 24"
                                                width="14"
                                                height="14"
                                                stroke="currentColor"
                                                strokeWidth="3"
                                                fill="none"
                                                strokeLinecap="round"
                                                strokeLinejoin="round"
                                                className={styles.closeButtonIcon}
                                            >
                                                <line x1="18" y1="6" x2="6" y2="18" />
                                                <line x1="6" y1="6" x2="18" y2="18" />
                                            </svg>
                                        }
                                        theme="inherit"
                                        onClick={onCloseClick}
                                    />
                                </div>

                                <div
                                    className={classNames({
                                        [styles.nubbin]: true,
                                        [styles.nubbinTop]: startsWith(placement, 'bottom'),
                                        [styles.nubbinBottom]: startsWith(placement, 'top'),
                                        [styles.nubbinLeft]: startsWith(placement, 'right'),
                                        [styles.nubbinRight]: startsWith(placement, 'left'),
                                    })}
                                    ref={arrowProps.ref}
                                    style={arrowProps.style}
                                />
                            </div>
                        )}
                    </Popper>
                )}
            </ConditionalPortal>
        </Manager>
    );
}

interface PopoverTitlePropTypes {
    /**
     * The contents of the title.
     */
    children: React.ReactNode;
}

const PopoverTitle = ({ children }: PopoverTitlePropTypes): JSX.Element => (
    <div className={styles.popoverTitle}>{children}</div>
);

interface PopoverBodyPropTypes {
    /**
     * The contents of the body.
     */
    children: React.ReactNode;
}

const PopoverBody = ({ children }: PopoverBodyPropTypes): JSX.Element => (
    <div className={styles.popoverBody}>{children}</div>
);

interface PopoverPrimaryButtonPropTypes {
    /**
     * The text of the button.
     */
    children: string;
    /**
     * The function to call when the button is clicked.
     */
    onClick: () => void;
}

const PopoverPrimaryButton = ({
    children,
    onClick,
}: PopoverPrimaryButtonPropTypes): JSX.Element => (
    <Themed size="small" onClick={onClick} theme="popover-primary">
        {children}
    </Themed>
);

interface PopoverSecondaryButtonPropTypes {
    /**
     * The text of the button.
     */
    children: string;
    /**
     * The function to call when the button is clicked.
     */
    onClick?: () => void;
    /**
     * Link to visit when the button is clicked. It will be opened in a new tab.
     */
    to?: string;
}

const PopoverSecondaryButton = ({
    children,
    onClick,
    to,
}: PopoverSecondaryButtonPropTypes): JSX.Element => (
    <Themed size="small" onClick={onClick} to={to} theme="popover-secondary" shouldOpenInNewTab>
        {children}
    </Themed>
);

export { PopoverTitle, PopoverBody, PopoverPrimaryButton, PopoverSecondaryButton };
