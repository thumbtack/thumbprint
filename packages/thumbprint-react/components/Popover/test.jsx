import React from 'react';
import { mount } from 'enzyme';
import Popover from './index';

// https://github.com/thumbtack/thumbprint/issues/72
// https://github.com/FezVrasta/popper.js/issues/478#issuecomment-341494703
jest.mock('popper.js', () => {
    const PopperJS = jest.requireActual('popper.js');

    class Popper {
        constructor() {
            return {
                destroy: () => {},
                scheduleUpdate: () => {},
            };
        }
    }

    Popper.placements = PopperJS.placements;

    return Popper;
});

test('renders a popover', () => {
    const onCloseClick = jest.fn();
    const wrapper = mount(
        <Popover isOpen onCloseClick={onCloseClick} launcher={() => null}>
            Goose
        </Popover>,
    );

    expect(wrapper).toMatchSnapshot();
});

test('initially traps focus to the first focusable element', () => {
    jest.useFakeTimers();

    const onCloseClick = jest.fn();
    const wrapper = mount(
        <Popover isOpen onCloseClick={onCloseClick} launcher={() => null}>
            <button type="button" id="hi">
                Hi
            </button>
        </Popover>,
    );

    // Run setTimeouts() in <FocusTrap> to completion
    jest.runAllTimers();

    const button = wrapper.find('#hi');
    expect(button.is(':focus')).toBe(true);

    jest.useRealTimers();
});
