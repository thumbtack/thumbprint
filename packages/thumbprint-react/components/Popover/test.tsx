import React from 'react';
import { mount } from 'enzyme';
import Popover from './index';

// https://github.com/thumbtack/thumbprint/issues/72
// https://github.com/FezVrasta/popper.js/issues/478#issuecomment-341494703
jest.mock('popper.js', (): object => {
    const PopperJS = jest.requireActual('popper.js');

    class Popper {
        public static placements: typeof PopperJS.placements;

        public constructor() {
            return {
                destroy: (): void => {},
                scheduleUpdate: (): void => {},
            };
        }
    }

    Popper.placements = PopperJS.placements;

    return Popper;
});

test('renders a popover', (): void => {
    const onCloseClick = jest.fn();
    const wrapper = mount(
        <Popover isOpen onCloseClick={onCloseClick} launcher={(): null => null}>
            Goose
        </Popover>,
    );

    expect(wrapper).toMatchSnapshot();
});

test('initially traps focus to wrapper, not the focusable child', (): void => {
    jest.useFakeTimers();

    const onCloseClick = jest.fn();
    const wrapper = mount(
        <Popover isOpen onCloseClick={onCloseClick} launcher={(): null => null}>
            <button type="button" id="hi">
                Hi
            </button>
        </Popover>,
    );

    // Run setTimeouts() in focus-trap to completion
    jest.runAllTimers();

    const modalWrapper = wrapper.find('[role="dialog"]');
    expect(modalWrapper.is(':focus')).toBe(true);

    const button = wrapper.find('#hi');
    expect(button.is(':focus')).toBe(false);

    jest.useRealTimers();
});
