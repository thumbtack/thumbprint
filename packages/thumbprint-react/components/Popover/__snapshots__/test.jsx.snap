// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders content that is passed in 1`] = `
<OnClickOutside(Popover)
  eventTypes={
    Array [
      "mousedown",
      "touchstart",
    ]
  }
  excludeScrollbar={false}
  isOpen={false}
  launcher={[Function]}
  onCloseClick={[Function]}
  outsideClickIgnoreClass="ignore-react-onclickoutside"
  preventDefault={false}
  stopPropagation={false}
>
  <Popover
    container="body"
    disableOnClickOutside={[Function]}
    enableOnClickOutside={[Function]}
    eventTypes={
      Array [
        "mousedown",
        "touchstart",
      ]
    }
    isOpen={false}
    launcher={[Function]}
    onCloseClick={[Function]}
    outsideClickIgnoreClass="ignore-react-onclickoutside"
    position="top"
    preventDefault={false}
    stopPropagation={false}
    wrappedRef={[Function]}
  >
    <Manager>
      <Reference>
        <InnerReference
          getReferenceRef={[Function]}
        />
      </Reference>
      <Portal
        containerInfo={
          <body>
            <div
              class="root"
              style="position: absolute; top: 0px; left: 0px; opacity: 0; pointer-events: none;"
            >
              Goose
              <div
                class="closeButton"
              >
                <button
                  aria-label="Close popover"
                  class="plain plainThemeInherit"
                  type="button"
                >
                  <span
                    class="flexCenter"
                  >
                    <svg
                      class="closeButtonIcon"
                      fill="currentColor"
                      height="14"
                      viewBox="0 0 14 14"
                      width="14"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M3.4346,2.24614 C3.10642,1.91795 2.57432,1.91795 2.24614,2.24614 C1.91795,2.57432 1.91795,3.10642 2.24614,3.4346 L5.8116,7.00006 L2.24627,10.5654 C1.91808,10.8936 1.91808,11.4257 2.24627,11.7539 C2.57445,12.082 3.10655,12.082 3.43473,11.7539 L7.00006,8.18853 L10.5654,11.7539 C10.8936,12.082 11.4257,12.082 11.7539,11.7539 C12.082,11.4257 12.082,10.8936 11.7539,10.5654 L8.18853,7.00006 L11.754,3.4346 C12.0822,3.10642 12.0822,2.57433 11.754,2.24614 C11.4258,1.91795 10.8937,1.91795 10.5655,2.24614 L7.00006,5.8116 L3.4346,2.24614 Z"
                      />
                    </svg>
                  </span>
                </button>
              </div>
              <div
                class="nubbin"
              />
            </div>
          </body>
        }
      >
        <PopoverContent
          isOpen={false}
          onCloseClick={[Function]}
          position="top"
        >
          <Popper
            modifiers={
              Object {
                "offset": Object {
                  "offset": "0, 8px",
                },
                "preventOverflow": Object {
                  "boundariesElement": "window",
                },
              }
            }
            placement="top"
            positionFixed={false}
          >
            <InnerPopper
              eventsEnabled={true}
              modifiers={
                Object {
                  "offset": Object {
                    "offset": "0, 8px",
                  },
                  "preventOverflow": Object {
                    "boundariesElement": "window",
                  },
                }
              }
              placement="top"
              positionFixed={false}
            >
              <div
                className="root"
                style={
                  Object {
                    "left": 0,
                    "opacity": 0,
                    "pointerEvents": "none",
                    "position": "absolute",
                    "top": 0,
                  }
                }
              >
                Goose
                <div
                  className="closeButton"
                >
                  <ForwardRef
                    accessibilityLabel="Close popover"
                    iconLeft={
                      <NavigationCloseTiny
                        className="closeButtonIcon"
                        fill="currentColor"
                        height="14"
                        viewBox="0 0 14 14"
                        width="14"
                        xmlns="http://www.w3.org/2000/svg"
                      />
                    }
                    isDisabled={false}
                    onClick={[Function]}
                    theme="inherit"
                    type="button"
                  >
                    <ForwardRef
                      accessibilityLabel="Close popover"
                      iconLeft={
                        <NavigationCloseTiny
                          className="closeButtonIcon"
                          fill="currentColor"
                          height="14"
                          viewBox="0 0 14 14"
                          width="14"
                          xmlns="http://www.w3.org/2000/svg"
                        />
                      }
                      isDisabled={false}
                      onClick={[Function]}
                      shouldOpenInNewTab={false}
                      theme="inherit"
                      to={null}
                      type="button"
                    >
                      <button
                        aria-label="Close popover"
                        className="plain plainThemeInherit"
                        disabled={false}
                        onClick={[Function]}
                        type="button"
                      >
                        <span
                          className="flexCenter"
                        >
                          <NavigationCloseTiny
                            className="closeButtonIcon"
                            fill="currentColor"
                            height="14"
                            viewBox="0 0 14 14"
                            width="14"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <svg
                              className="closeButtonIcon"
                              fill="currentColor"
                              height="14"
                              viewBox="0 0 14 14"
                              width="14"
                              xmlns="http://www.w3.org/2000/svg"
                            >
                              <path
                                d="M3.4346,2.24614 C3.10642,1.91795 2.57432,1.91795 2.24614,2.24614 C1.91795,2.57432 1.91795,3.10642 2.24614,3.4346 L5.8116,7.00006 L2.24627,10.5654 C1.91808,10.8936 1.91808,11.4257 2.24627,11.7539 C2.57445,12.082 3.10655,12.082 3.43473,11.7539 L7.00006,8.18853 L10.5654,11.7539 C10.8936,12.082 11.4257,12.082 11.7539,11.7539 C12.082,11.4257 12.082,10.8936 11.7539,10.5654 L8.18853,7.00006 L11.754,3.4346 C12.0822,3.10642 12.0822,2.57433 11.754,2.24614 C11.4258,1.91795 10.8937,1.91795 10.5655,2.24614 L7.00006,5.8116 L3.4346,2.24614 Z"
                              />
                            </svg>
                          </NavigationCloseTiny>
                        </span>
                      </button>
                    </ForwardRef>
                  </ForwardRef>
                </div>
                <div
                  className="nubbin"
                  style={Object {}}
                />
              </div>
            </InnerPopper>
          </Popper>
        </PopoverContent>
      </Portal>
    </Manager>
  </Popover>
</OnClickOutside(Popover)>
`;
